api/internalutils/stream/stream.go

```diff
@@ -0,0 +1,314 @@
+// represent the complete set.
+func CollectPages[T any](stream Stream[[]T]) ([]T, error) {
+	var c []T
+	for stream.Next() {
+		c = append(c, stream.Item()...)
+	}
+	return c, trace.Wrap(stream.Done())
+}
+
+// filterMap is a stream that performs a FilterMap operation.
+type filterMap[A, B any] struct {
+	inner Stream[A]
+	fn    func(A) (B, bool)
+	item  B
+}
+
+func (stream *filterMap[A, B]) Next() bool {
+	for {
+		if !stream.inner.Next() {
+			return false
+		}
+		var ok bool
+		stream.item, ok = stream.fn(stream.inner.Item())
+		if !ok {
+			continue
+		}
+		return true
+	}
+}
+
+func (stream *filterMap[A, B]) Item() B {
+	return stream.item
+}
+
+func (stream *filterMap[A, B]) Done() error {
+	return stream.inner.Done()
+}
+
+// FilterMap maps a stream of type A into a stream of type B, filtering out
+// items when fn returns false.
+func FilterMap[A, B any](stream Stream[A], fn func(A) (B, bool)) Stream[B] {
+	return &filterMap[A, B]{
+		inner: stream,
+		fn:    fn,
+	}
+}
+
+// mapWhile is a stream that performs a MapWhile operation.
+type mapWhile[A, B any] struct {
+	inner Stream[A]
+	fn    func(A) (B, bool)
+	item  B
+}
+
+func (stream *mapWhile[A, B]) Next() bool {
+	if !stream.inner.Next() {
+		return false
+	}
+
+	var ok bool
+	stream.item, ok = stream.fn(stream.inner.Item())
+	return ok
+}
+
+func (stream *mapWhile[A, B]) Item() B {
+	return stream.item
+}
+
+func (stream *mapWhile[A, B]) Done() error {
+	return stream.inner.Done()
+}
+
+// MapWhile maps a stream of type A into a stream of type B, halting early
+// if fn returns false.
+func MapWhile[A, B any](stream Stream[A], fn func(A) (B, bool)) Stream[B] {
+	return &mapWhile[A, B]{
+		inner: stream,
+		fn:    fn,
+	}
+}
+
+// empty is a stream that halts immediately
+type empty[T any] struct {
+	err error
+}
+
+func (stream empty[T]) Next() bool {
+	return false
+}
+
+func (stream empty[T]) Item() T {
+	panic("Item() called on empty/failed stream")
+}
+
+func (stream empty[T]) Done() error {
+	return stream.err
+}
+
+// Fail creates an empty stream that fails immediately with the supplied error.
+func Fail[T any](err error) Stream[T] {
+	return empty[T]{err}
+}
+
+// Empty creates an empty stream (equivalent to Fail(nil)).
+func Empty[T any]() Stream[T] {
+	return empty[T]{}
+}
+
+// once is a stream that yields a single item
+type once[T any] struct {
+	yielded bool
+	item    T
+}
+
+func (stream *once[T]) Next() bool {
+	if stream.yielded {
+		return false
+	}
+	stream.yielded = true
+	return true
+}
+
+func (stream *once[T]) Item() T {
+	return stream.item
+}
+
+func (stream *once[T]) Done() error {
+	return nil
+}
+
+// Once creates a stream that yields a single item.
+func Once[T any](item T) Stream[T] {
+	return &once[T]{
+		item: item,
+	}
+}
+
+// Drain consumes a stream to completion.
+func Drain[T any](stream Stream[T]) error {
+	for stream.Next() {
+	}
+	return trace.Wrap(stream.Done())
+}
+
+// slice streams the elements of a slice
+type slice[T any] struct {
```

Comment:
Can this be optimized to not hold an extra `int`?

Reply:
I think we're always going to need *some* kind of extra state.  If we just track our place by truncating the slice, then we'd skip the first item, because calls to `Next()` happen before calls to `Item()`.  Current impl initialzies the index to `-1` in order to handle this problem.  We could use a bool to make the first call to `Next()` a special case, but I don't see much benefit to that.